### ISSUE TITLE
Use backrest as a single user at home

### ISSUE THREAD

**Issue by stefnotch** [garethgeorge/backrest#283]
Note: if you have a question or want discussion please post in the [discussions area](https://github.com/garethgeorge/backrest/discussions).

**Is your feature request related to a problem? Please describe.**

I would like to be able to recommend backup programs, like restic/rustic to non-techy family members. This means that a GUI, like what backrest offers, is awesome.

However, it currently seems to not be designed to be a program where I can simply
1. Download it
2. Start it - automatically opens a GUI window
3. Do whatever I need
4. Close it

Instead, it is
1. Download it
2. Start it, and be surprised about a CLI window
3. Navigate to http://localhost:9898/ in the browser
4. Do whatever I need
5. Close the CLI window

**Describe the solution you'd like**

Would it be reasonable to offer a variant of the GUI which comes with an embedded web browser? (e.g. Tauri or Electron)
Then, it'd be possible for backrest to work like

1. Download it
2. Start it, backrest and a HTTP server get started in the background, and a normal looking GUI gets opened. That normal looking GUI is actually a fully fledged web browser. In the case of Electron, it's a massive embedded Chrome, in the case of Tauri it elegantly embeds the system's web browser.
3. I do whatever I need
4. I close the window, which also gracefully shuts down the HTTP server and backrest.

**Comment by stefnotch** [comment]
Apparently there are webview libraries for Go, such as https://github.com/webview/webview_go and https://github.com/pojntfx/hydrapp . There are also a few others which either seem to be unmaintained, or which don't support important browsers.

**Comment by garethgeorge** [comment]
Hey, just to clarify a bit -- is this a problem you're running into on Windows? An open issue is that Backrest's windows operations aren't very streamlined. 

I'd happily accept community contributions to improve the state of things here, my windows knowledge isn't particular deep so it's not something I've been working on. Ideally Backrest _should_ include an installer that runs the binary as a background service.

On MacOS / Linux the installer should already be setup to run the binary in the background (no visible window). 

On the GUI front, this is another place where I'd happily accept community work to shim an electron app around Backrest as I can definitely understand the utility but probably it isn't a focus for me and isn't something I'm interested to maintain at the moment.

**Comment by stefnotch** [comment]
Yes, I'm running into this on Windows. If my target audience were tech-y enough to use Linux, then I'd just show them the restic CLI.

I suggested the webview option, because if such a library is mature enough, then using it is relatively simple, see https://github.com/webview/webview_go/blob/master/examples/basic/main.go . 
Of course that doesn't cover other parts that one might want, like a tray icon.

**Comment by aep** [comment]
I'm interested in this for commercial reasons so we can definitely help with contributions if there's a consensus.

I don't think an "native app" really is required tho. In my opinion it's completely fine to open the default Webbrowser. 

What's more important is a desktop shortcut that does that.

**Comment by stefnotch** [comment]
@aep I've gotten feedback from multiple people that apps which just open the webbrowser are somewhat disliked. They pretty much need a tray icon to tell the user that the app is still running, and to let the user close the app.
It also looks very confusing when the user closes the app, but the website is still open. So one has to get the website to regularly check if the app is still running, and if not, the website has to show some sort of message. 
And then there's the tidbit about having to support weirder configurations, like a user who has NoScript or an overzealous adblocker installed.

Hence my request to use a webview library. They open an already installed and webbrowser as a separated window, with a separate profile, which makes them very lightweight and shockingly easy to integrate. And they avoid all the annoyances mentioned above.

**Comment by garethgeorge** [comment]
Hey all, just chiming in to say that community contributions re: a windows installer would be very welcome.

I think the right place to put the code for any contributions would be in a `build` subfolder (following the rough example here https://github.com/golang-standards/project-layout/tree/master/build) and I should probably consolidate the Dockerfiles and other release related code in there.

There's a bug tracking an installer https://github.com/garethgeorge/backrest/issues/163 but I don't think any recent progress has been made. Ideally we are able to offer a simple binary that we can package with the release (given a PR adding a codegen for a windows installer I can easily add it to the release workflow). 

I found https://github.com/getlantern/systray which adds a system tray icon to a windows binary -- there's an interesting question of whether we should reconfigure the build to make this the default OR if we should ship two versions of the binary (e.g. a standalone and perhaps a system tray binary that is bundled with an msi installer?). It's probably OK to just provide the msi installer + tray binary so long as we still provide a way to access the process logs on windows.

**Comment by sebcej** [comment]
Another great possibility can be [Wails](https://github.com/wailsapp/wails). I'm thinking about making a try in the near future

**Comment by garethgeorge** [comment]
> @garethgeorge by far the easiest way to solve this is with [nsis](https://nsis.sourceforge.io/Main_Page), I created a simple NSIS script for this, I think it's the best solution because it also can be compiled cross platform, as there are nsis compilers for linux.

This looks great, thanks for putting that install script (and icon) together! I can run forward with this to setup an installer for windows. I agree that I need to create a new windows build that doesn't open a console. I'll take a look at using https://github.com/getlantern/systray to provide a separate skew of the binary for windows that both opens a system tray icon to make it visible that it's running (and additionally hides it's console window).

> Another great possibility can be [Wails](https://github.com/wailsapp/wails). I'm thinking about making a try in the near future

I'd be very interested to hear how that goes. I think Backrest should operate as a daemon irrespective of the GUI but a Wails GUI as a client for the daemon makes sense to me.

**Comment by garethgeorge** [comment]
Started on integrating the installer and setting up a tray application in https://github.com/garethgeorge/backrest/pull/294/files. Using the tray app to fork backrest with the right options solves the console window problem.

**Comment by garethgeorge** [comment]
Merged with a working first pass, the solution I landed on is 

 * backrest-windows-tray binary which is essentially a process host for backrest. The tray binary hosts the windows tray widget which allows you to open the web UI, open logs, or stop backrest. It also forks the backrest executable and pipes its logs to a log directory.
 * installer installs both backrest and backrest-windows-tray, backrest-windows-tray is the binary that is configured to autostart.

Tested the x86 binary and working nicely on my windows machine (I don't have an arm64 machine to test the other builds), preview binaries available on the CI/CD runs for the merged run: 

 * https://github.com/garethgeorge/backrest/actions/runs/9231407309 
 * https://github.com/garethgeorge/backrest/actions/runs/9231407309/artifacts/1536676724

**Comment by ThaUnknown** [comment]
@garethgeorge I noticed there's an issue in the build script i provided, all the icon shortcuts are wrong, instead of `"${BUILD_DIR}\icon.ico"` they should be `"$INSTDIR\icon.ico"`

**Comment by garethgeorge** [comment]
No problem, it was a really good start point & worked with a few edits. I found and fixed that, thanks for calling it out. I did do some debugging / wrangling of the script to make it compatible with the setup the ci/cd pipeline needs.

**Comment by PovilasID** [comment]
1st I really like this project but to recommend it to mostly windows user there a couple considerations.
1. Single .exe installer. It can be a web wrapper like suggested that is fine.
2. No open ports. I would prefer user who do not know much to not have an open port on their device. I get that it is not accessible form the local machine only but I would feel much better if client machine for backup was not exposing the UI to the ethernet protocol at all. 
3. Reasonable defaults. Now in new Plan UI retention defaults are set to 0 and there are no planned backups. I just do not want to give user more opportunities to fail to setup basics.
4. Launch on startup. This ties in the installer but it should really add it to startup queue, so user do not forget to launch it.

**Comment by Ambient-Impact** [comment]
> Fully agree on the systray. That's a thing windows users expect.
> 
> Not sure about the wrapper app yet, but if someone else wants to pick that up why not. Currently considering if we can help with the installer first.

Agreed on both counts, and I'd add even on Linux a system tray icon is appreciated for those of us using a desktop manager that supports it - e.g. KDE Plasma, GNOME, etc.

**Comment by Ambient-Impact** [comment]
> I'm interested in this for commercial reasons so we can definitely help with contributions if there's a consensus.
> 
> I don't think an "native app" really is required tho. In my opinion it's completely fine to open the default Webbrowser.
> 
> What's more important is a desktop shortcut that does that.

It's a chicken and egg problem: if apps keep reinforcing that only native apps are serious/trustworthy/etc. by de-emphasizing web UIs or making them secondary to native apps, users will continue to see them as second class; on the other hand, if enough apps push for web UIs to normalise that as equal to native apps, perceptions will change.

**Comment by Ambient-Impact** [comment]
With the massive caveat that I'm a web and not native developer, I'd also recommend Neutron as an alternative to Electron if you want help chip away at Google's near monopoly: https://github.com/gamingdoom/neutron

**Comment by garethgeorge** [comment]
To add on to this -- agreed with all of the points @Ambient-Impact impact makes :) . I think Backrest aims to stay a webUI as it's by far the easiest way to provide a good cross platform experience & focus on feature work instead of issues with platform ports.

Happy to discuss packaging if anyone is interested in porting the tray app to new platforms, it's not a priority on my road map at the moment but agree that it would be very nice to have.
